runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/stream/KafkaClientSaslHandshaker.java

```diff
@@ -430,8 +568,98 @@ protected final int decodeSaslAuthenticate(
+                {
+                    DirectBuffer serverFirstResponse = authenticateResponse.authBytes().value();
+                    String serverFirstMessage = serverFirstResponse.getStringWithoutLengthUtf8(0,
+                            serverFirstResponse.capacity());
+
+                    Matcher serverScramResponse = SERVER_FIRST_MESSAGE.matcher(serverFirstMessage);
+                    while (serverScramResponse.find())
+                    {
+                        client.serverNonce = serverScramResponse.group(1);
+                        client.salt = serverScramResponse.group(2);
+                        client.iterationCount = serverScramResponse.group(3);
+                    }
+                    if (errorCode ==  0 && client.serverNonce.startsWith(client.clientNonce))
+                    {
+                        client.onDecodeSaslResponse(traceId);
+                        client.doEncodeSaslScramAuthenticateRequest(traceId, budgetId);
+                    }
+                    else
+                    {
+                        client.onDecodeSaslResponse(traceId);
+                        client.onDecodeSaslAuthenticateResponse(traceId, authorization, 58);
+                    }
+                }
+                else
+                {
+                    client.onDecodeSaslResponse(traceId);
+                    client.onDecodeSaslAuthenticateResponse(traceId, authorization, errorCode);
+                }
+            }
+        }
+
+        return progress;
+    }
+
+    protected final int decodeSaslScramAuthenticate(
+            KafkaSaslClient client,
+            long traceId,
+            long authorization,
+            long budgetId,
+            int reserved,
+            DirectBuffer buffer,
+            int offset,
+            int progress,
+            int limit)
+    {
+        final int length = limit - progress;
+
+        if (length != 0)
+        {
+            final SaslAuthenticateResponseFW authenticateResponse = saslAuthenticateResponseRO.tryWrap(buffer, progress, limit);
+
+            if (authenticateResponse != null)
+            {
+                final int errorCode = authenticateResponse.errorCode();
+
+                progress = authenticateResponse.limit();
+
+                client.decodableResponseBytes -= authenticateResponse.sizeof();
+                assert client.decodableResponseBytes >= 0;
+
+                try
+                {
+                    byte[] serverKey = formatter.serverKey(client.saltedPassword);
+                    byte[] serverSignature = formatter.hmac(serverKey, client.authMessage);
+                    DirectBuffer serverFinalResponse = authenticateResponse.authBytes().value();
+                    String serverFinalMessage = serverFinalResponse.getStringWithoutLengthUtf8(0,
+                            serverFinalResponse.capacity());
+
+                    Matcher serverFinalResponseMatcher = SERVER_FINAL_MESSAGE.matcher(serverFinalMessage);
```

Comment:
Anywhere you need to use a `Matcher`, pre-create a non-static instance at the `KafkaClientSaslHandshaker` scope by calling `.matcher("")`, and then use `.reset(serverFinalMessage)` to reuse instead of creating a new one each time.